### PR TITLE
Hw(lecture 4).oop

### PR DESCRIPTION
**Pull request by olgakuzmins**
Tasks 1-6 are done:
1. added abstract class BasicEntity
2. added interface Printable
3. added getters and setters for Ticket's fields
4. added method share() to TicketService
5. added User, Client and Admin
6. overriden equals(), hashcode() and toString() in Ticket

### CODE REVIEW

File: src/model/BasicEntity.java [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 import java.util.UUID;
 
-public abstract class BasicEntity {
+public abstract class BasicEntity implements Printable {
     protected UUID id = UUID.randomUUID();
```
Comment (AtticSiol) [inline src/model/BasicEntity.java]: move setting to constuctor

File: src/model/Printable.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package model;
+
+public interface Printable {
+
+    default void print() {
+        System.out.println(this);
```
Comment (AtticSiol) [inline src/model/Printable.java]: add some info

File: src/model/Ticket.java [hunk 3]
```diff
@@ -74,17 +116,38 @@ public void checkEventCode(String eventCode) {
...
+                "eventCode = " + eventCode + '\n' +
+                "time = " + formatDate(time) + '\n' +
+                "isPromo = " + isPromo + '\n' +
+                "sector = " + sector + '\n' +
+                "backpackWeight = " + backpackWeight + '\n' +
+                "ticketCreationTime = " + formatDate(ticketCreationTime) + '\n' +
+                "price = " + formatter.format(price) + '\n';
+    }
+
+    private String formatDate(Instant instant) {
```
Comment (AtticSiol) [inline src/model/Ticket.java:139]: move in the end of the class

File: src/model/Ticket.java [hunk 4]
```diff
@@ -74,17 +116,38 @@ public void checkEventCode(String eventCode) {
 
     @Override
     public String toString() {
```
Comment (AtticSiol) [inline src/model/Ticket.java:127]: move after equals and hashCode

File: src/service/TicketService.java [hunk 5]
```diff
@@ -77,9 +76,65 @@ public List<Ticket> returnTicketsBySector(Sector sector) {
         return sectorTickets;
     }
 
+    public void shareTicket(Ticket ticket, String phoneNumber) {
+        System.out.println("The ticket " + checkTicket(ticket).getId() + " is sent to the number " + checkPhoneNumber(phoneNumber));
+    }
+
+    public void shareTicket(Ticket ticket, String phoneNumber, String email) {
+        System.out.println("The ticket " + checkTicket(ticket).getId() + " is sent to the number " + checkPhoneNumber(phoneNumber) + " and email " + checkEmail(email));
+    }
```
Comment (AtticSiol) [inline src/service/TicketService.java]: extract it to interface

File: src/model/Ticket.java [hunk 6]
```diff
@@ -24,7 +30,7 @@ public Ticket(String concertHall, String eventCode, Instant time) {
         this.time = time;
     }
 
-    public Ticket(String id, String concertHall,
+    public Ticket(UUID id, String concertHall,
```
Comment (AtticSiol) [inline src/model/Ticket.java]: You do not need to use "UUID id" here as argument. this field will fill with generated value from constructor in BaseEntity

Comment (olgakuzmins) [reply src/model/Ticket.java]: Yes, sure. I left it in case automatical generated id will have been be a wrong solution for the project.

Comment (olgakuzmins) [reply src/model/Ticket.java]: So, I deleted this extra "UUID id"

Comment (AtticSiol) [reply src/model/Ticket.java]: no, it is good solution. approved

**Review by AtticSiol — APPROVED** [review]

**Pull request merged**